src/Transport/Sending/MessageDispatcher.cs

```diff
@@ -10,108 +12,205 @@
-                        scope.Complete();
-                    }
-                }
-                finally
+                var messagesToSend = new Queue<ServiceBusMessage>(operations.Count);
+                foreach (var operation in operations)
                 {
-                    messageSenderPool.ReturnMessageSender(sender, client);
+                    var message = operation.Message.ToAzureServiceBusMessage(operation.Properties, partitionKey);
+                    operation.ApplyCustomizationToOutgoingNativeMessage(message, transaction, Log);
+                    messagesToSend.Enqueue(message);
                 }
+                dispatchTasks.Add(DispatchBatchForDestination(destination, client, committableTransaction, messagesToSend, cancellationToken));
             }
+        }
 
-            foreach (var transportOperation in multicastTransportOperations)
+        async Task DispatchBatchForDestination(string destination, ServiceBusClient client, CommittableTransaction committableTransaction, Queue<ServiceBusMessage> messagesToSend, CancellationToken cancellationToken)
+        {
+            var messageCount = messagesToSend.Count;
+            int batchCount = 0;
+            var sender = messageSenderRegistry.GetMessageSender(destination, client);
+            while (messagesToSend.Count > 0)
             {
-                var sender = messageSenderPool.GetMessageSender(topicName, client);
+                using ServiceBusMessageBatch messageBatch = await sender.CreateMessageBatchAsync(cancellationToken)
+                    .ConfigureAwait(false);
 
-                try
+                StringBuilder logBuilder = null;
+                if (Log.IsDebugEnabled)
                 {
-                    var message = transportOperation.Message.ToAzureServiceBusMessage(transportOperation.Properties, partitionKey);
+                    logBuilder = new StringBuilder();
+                }
 
-                    ApplyCustomizationToOutgoingNativeMessage(transportOperation, message, transaction);
+                var peekedMessage = messagesToSend.Peek();
+                if (messageBatch.TryAddMessage(peekedMessage))
+                {
+                    var added = messagesToSend.Dequeue();
+                    if (Log.IsDebugEnabled)
+                    {
+                        added.ApplicationProperties.TryGetValue(Headers.MessageId, out var messageId);
+                        logBuilder!.Append($"{messageId ?? added.MessageId},");
+                    }
+                }
+                else
+                {
+                    peekedMessage.ApplicationProperties.TryGetValue(Headers.MessageId, out var messageId);
+                    throw new ServiceBusException($"Message #{messageCount - messagesToSend.Count} with message id '{messageId ?? peekedMessage.MessageId}' is too large and cannot be sent.", ServiceBusFailureReason.MessageSizeExceeded);
```

Comment:
>"Message #N with message id 'XXXXXXXXXXX' is too large and cannot be sent."

That's not necessarily accurate or helpful.

Realistically, thinking about why it could happen, there are 2 common scenarios:
1. Too many messages combined together exceeded the maximum and
2. One of the messages was big.

It would help to throw an exception with messages attempted for dispatch as a batch
- Total size of the 
- Message types involved in the batch

Reply:
@SeanFeldman thanks for the feedback. You are not late to the party. We are planning to do a few more rounds of improvements. I will send in a PR that also addresses your input :crossed_fingers:

Reply:
@SeanFeldman I was thinking about this a bit more and I think there is not much we can do without significantly impacting the success path as well as reimplementing parts of the logic that is ultimately part of the batch and should be owned by the SDK. 

Because we are building up the batch by peeking we would need to store all the information that might be helpful to be included in the case when a batch fails proactively into some structure that is only ever used when the `TryAdd` returned `false`. This is because the batch itself doesn't allow to be enumerated over the content it owns. In order to get the total size of the message we would need to do a back of the napkin calculation in this case too. The benefit of the batch as it stands today is that it encapsulates logic of the service bus tier, message length calculations and more which we would have to duplicate up again into the code. 

I will try to see if I can make the exception message slightly more helpful but going down the path of storing a lot of additional contextual information plus re-implementing parts of the SDK for the failure case is not what we would like to do. 

Can you maybe think of an acceptable tradeoff? FYI I'm not rejecting the input. I will continue to think about better approaches.

Reply:
Makes sense. In that case, something that would guide towards troubleshooting w/o leading inna direction that won't be much of a help. Like the message #. Perhaps a big blob of text explaining how this could have happened, to guide the users towards fixing what only they can fix. E.g.
"The messages attempted for dispatch have either collectively exceeded the maximum batch size N or one of the messages has the maximum size of N. Review your code, reduce messages size by using data bus, or use the Premium Service Bus tier." Something amongst those lines.

Reply:
What do you think about this https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/pull/649 ?